README
NAME
    ReRe - Simple Redis Rest Interface

VERSION
    version 0.021

SYNOPSIS
  Console
        # authentication by ip address.
        curl -L http://127.0.0.1:3000/redis/set/foo/bar
        {"set":{"foo":"bar"}}

        curl -L http://127.0.0.1:3000/redis/get/foo
        {"get":{"foo":"bar"}}

        curl -L http://127.0.0.1:3000/redis/info
        {"info":{"last_save_time":"1306939038", ... }}

        # authentication by http basic.
        curl --basic -u 'userro:userro' -L http://127.0.0.1:3000/redis/get/foo
        {"get":{"foo":"bar"}}

        # JSONP Callback.
        curl --basic -u 'userro:userro' -L
        http://127.0.0.1:3000/redis/get/foo\?callback=MyCB
        MyCB( {"get":{"foo":"bar"}} )

  Perl
        use ReRe::Client;
        my $conn = ReRe::Client->new(
            {   url      => '127.0.0.1:3000',
                username => 'userro',
                password => 'userro'
            } );

        $conn->set( 'foo', '2' );

        $conn->get('foo')

    See ReRe::Client for more information.

DESCRIPTION
    ReRe is a simple redis rest interface write in Perl and Mojolicious,
    with some features like:

  What is ReRe ?
    Access your Redis database directly from Javascript.
    Allows you to plugin other technologies: caching via varnish, proxying
    via HAProxy, authentication for APIs, etc. because HTTP is well
    supported.
    Access control list for methods of redis ;
    Support to run as daemon (simple web-server), CGI, FastCGI or PSGI ;
    HTTP Basic Auth security
    Support for GET and POST.
    Support for JSONP.
    HTTP 1.1 pipelining (fastcgi and psgi)
    CIDR authentication
    Bad support for Websockets
    Hooking, you can alter the behavior of an request method.
    ReRe::Client for write client applications in Perl.
    REST interface to make your life easy in some world ;
    Simple to install and use

  What is not ReRe ?
    Please, don't use this application if ...

    If you are looking for a local database for cache or key-value
    If you are looking for performance.

Configuration
  users.conf
        <users>
            <userro>
                password userro
                roles get
            </userro>

            <userrw>
                password userrw
                roles get set info
            </userrw>

            <userall>
                allow 127.0.0.1
                password userall
                roles all
            </userall>

            <userlocalnet>
                allow 192.168.0.0/24
                roles get info
            </userlocalnet>
        </users>

  server.conf
        <server>
            host 127.0.0.1
            port 6379
            hooks Log
        </server>

Hooks
    You can alter the behaivor of any method or create your own.

  Example
        package ReRe::Hook::MyHook;
        use Moose::Role;

        sub _hook {
            my $self = shift;

            my $method = $self->method;
            my $args = $self->args;
            my $conn = $self->conn;

            if ( $method eq 'set' ) {
                my ( $key, $value ) = @{$args};

                if ( $key eq 'immutable' ) {
                    return { err => 'this object is immutable' }
                }

                if ( $key eq 'semaphoro' ) {
                    my $new_value = 0;
                    $new_value = 1 if $value eq 'red';
                    return $self->conn->set($key, $new_value)
                }
            }

        return 0; # if you want to process the origin method.
        }

        1;

Documentation
    More information, you can read in <http://www.rere.com.br>.

Development
    ReRe is a open source project for everyone to participate. The code
    repository is located on github. Feel free to send a bug report, a pull
    request, or a beer.

    <http://www.github.com/maluco/ReRe>

METHODS
  start
    Start ReRe.

  process
    Process the request to redis server.

SUPPORT
    You can find documentation for this module with the perldoc command.

    perldoc ReRe perldoc ReRe::Config perldoc ReRe::Server perldoc
    ReRe::User

    You can also look for information at:

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=ReRe>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/ReRe>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/ReRe>

    *   Search CPAN

        <http://search.cpan.org/dist/ReRe>

AUTHOR
    Thiago Rondon <thiago@nsms.com.br>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2011 by Thiago Rondon.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



CHANGES
Revision History for ReRe
=========================

0.020
-----

commit 7e37c0070e8dad36e02aa9003e2d25f4cddf12bf
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Thu Jun 9 18:53:30 2011 -0300

    dzil make Changes from git-log.
    
    Signed-off-by: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

commit 0edee0f57660310743e0a06fdcd735234d53fd72
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Thu Jun 9 18:31:56 2011 -0300

    bugfix with 0 value.
    
    Signed-off-by: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

v0.019 2011-06-08
-----------------

commit f4e6bf0e641c374f52e41c5b8520e8ebd4525a88
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Wed Jun 8 18:29:14 2011 -0300

    v0.019

commit a4f9a4645906c65f4cae15dc9486f844c9004469
Author: Gabriel A. Santana <gabiruh@gmail.com>
Date:   Wed Jun 8 16:57:31 2011 -0300

    - using MooseX::SimpleConfig
    - attempt to improve ReRe::Hook achitecture
    
    Signed-off-by: Gabriel A. Santana <gabiruh@gmail.com>

commit 6936458da3cc45b0071fdac8b910727e5566be95
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Wed Jun 8 10:17:51 2011 -0300

    more tests..
    
    Signed-off-by: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

v0.018 2011-06-07
-----------------

commit 78d8df8348aecf590c72abf9d2dca9251205b5a3
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Tue Jun 7 21:44:18 2011 -0300

    Hmmm (...) .
    
    Signed-off-by: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

commit 694d15bbf25f0881142bdf03cd9465861a0a877e
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Tue Jun 7 11:40:02 2011 -0300

    Tests for ReRe::Client.
    
    Signed-off-by: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

v0.017 2011-06-07
-----------------

commit 49d5142f8eac2fc5bd1a1cafc15a286b95d5a9d9
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Tue Jun 7 00:49:08 2011 -0300

    Thanks Dist::Zilla.
    
    Signed-off-by: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

commit 20c6b0a27844a0a912a4b90427c0956af69afe79
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Tue Jun 7 00:13:12 2011 -0300

    JSONP support.
    
    Signed-off-by: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

commit e94d7977483ffb8992958b3d35a6f2a7312027a6
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Mon Jun 6 23:11:43 2011 -0300

    Websocket package.
    
    Signed-off-by: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

commit 198da19c196d1d9c313e2cf23682e37e0bad741a
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Mon Jun 6 19:43:28 2011 -0300

    return 0 in Log.pm
    
    Signed-off-by: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

commit f406cd7982bdb26255e9177f9dd2561f45030a84
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Mon Jun 6 19:42:01 2011 -0300

    Return option in hooks.
    
    Signed-off-by: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

commit e5d9068fec2cb71ede558b8db6744d048939ebdd
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Mon Jun 6 19:38:04 2011 -0300

    Config for websockets.
    
    Signed-off-by: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

commit 29af66b4ffee5bbd5333faadf9a0eea6bb29c1b1
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Mon Jun 6 17:41:44 2011 -0300

    add support for password in server.
    
    Signed-off-by: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

v0.016 2011-06-06
-----------------

commit f3d601319f8b9048a740207f91fb8f12f2d20e2e
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Mon Jun 6 16:10:24 2011 -0300

    typo.
    
    Signed-off-by: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

commit 0f113c53bbab8fcf94fb829089c34b908b47cd75
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Mon Jun 6 16:05:34 2011 -0300

    Go, 0.016.
    
    Signed-off-by: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

commit 2cc4f4e66dbc60ed03d9249155f66de07aa3e53a
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Mon Jun 6 15:39:29 2011 -0300

    perltidy and prepare new release.
    
    Signed-off-by: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

commit 20a61ff5a3c5a707588944b557557ad8348b7674
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Mon Jun 6 15:17:12 2011 -0300

    Add test for hooks.
    
    Signed-off-by: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

commit 3145165527590767970858c0e3826390f5115564
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Mon Jun 6 15:16:38 2011 -0300

    Add conn object to hooks.
    
    Signed-off-by: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

v0.015 2011-06-06
-----------------

commit 660c08a9e905332912cac11427a9fe4de09e9672
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Mon Jun 6 01:07:53 2011 -0300

    Ooops, hook files.
    
    Signed-off-by: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

commit 6bb5ce8d36d0f87c8e48c304b6b51ea871f3cb24
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Mon Jun 6 00:58:00 2011 -0300

    Simple support for hooks.
    
    Signed-off-by: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

v0.014 2011-06-05
-----------------

commit 1334cda3cbe293e818798925e924b90a79448819
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Sun Jun 5 18:26:52 2011 -0300

    [Git::Push]
    
    Signed-off-by: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

commit d5a637e3d1bf5b8bdd13489d7c67040da5e40027
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Sun Jun 5 18:24:46 2011 -0300

    eol
    
    Signed-off-by: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

commit 65d33147e92c6278b22cff1d18ab787574e647f8
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>
Date:   Sun Jun 5 18:17:06 2011 -0300

    Workround in websocket example.
    

lib/ReRe.pm

package ReRe;

use Moose;
use ReRe::User;
use ReRe::Server;
use ReRe::Websocket;

use List::Util qw(first);

# ABSTRACT: Simple Redis Rest Interface
our $VERSION = '0.021'; # VERSION

for my $item (qw/users server websocket/) {
  has "config_$item" => (
    is      => 'rw',
    isa     => 'Str',
    default => sub {
      my $options = [ "/etc/rere/$item.conf", "etc/$item.conf" ];
      my $found = first { -r } @$options;
      return $found if $found;
      die qq{Couldn't find a config file for $item, tried: } . join( ', ', @$options );
    }
  );
}

has user => (
  is   => 'ro',
  isa  => 'ReRe::User',
  lazy => 1,
  default =>
    sub { ReRe::User->new_with_config( configfile => shift->config_users ) }
);

has server => (
  is         => 'rw',
  isa        => 'ReRe::Server',
  lazy_build => 1,
  predicate  => 'has_server',
);

has websocket => (
  is      => 'rw',
  isa     => 'ReRe::Websocket',
  lazy    => 1,
  default => sub {
    ReRe::Websocket->new_with_config( configfile => shift->config_websocket );
  }
);

sub _build_server {
  my $self = shift;
  return ReRe::Server->new_with_config( configfile => $self->config_server );

}


sub start {
  my $self = shift;
  $self->user->process;
  $self->server;
}


sub process {
  my $self     = shift;
  my $username = shift;
  my $method   = shift;
  my @args     = @_;

  return { err => 'no_permission' }
    unless $self->user->has_role( $username, $method );

  my $ret = $self->server->execute( $method, @args );

  return { $method => [ @{$ret} ] } if ref($ret) eq 'ARRAY';
  return { $method => $ret };
}


1;


__END__
=pod

=head1 NAME

ReRe - Simple Redis Rest Interface

=head1 VERSION

version 0.021

=head1 SYNOPSIS

=head2 Console

    # authentication by ip address.
    curl -L http://127.0.0.1:3000/redis/set/foo/bar
    {"set":{"foo":"bar"}}

    curl -L http://127.0.0.1:3000/redis/get/foo
    {"get":{"foo":"bar"}}

    curl -L http://127.0.0.1:3000/redis/info
    {"info":{"last_save_time":"1306939038", ... }}

    # authentication by http basic.
    curl --basic -u 'userro:userro' -L http://127.0.0.1:3000/redis/get/foo
    {"get":{"foo":"bar"}}

    # JSONP Callback.
    curl --basic -u 'userro:userro' -L
    http://127.0.0.1:3000/redis/get/foo\?callback=MyCB
    MyCB( {"get":{"foo":"bar"}} )

=head2 Perl

    use ReRe::Client;
    my $conn = ReRe::Client->new(
        {   url      => '127.0.0.1:3000',
            username => 'userro',
            password => 'userro'
        } );

    $conn->set( 'foo', '2' );

    $conn->get('foo')

See L<ReRe::Client> for more information.

=head1 DESCRIPTION

ReRe is a simple redis rest interface write in Perl and L<Mojolicious>,
with some features like:

=head2 What is ReRe ?

=over 4

=item Access your Redis database directly from Javascript.

=item Allows you to plugin other technologies: caching via varnish, proxying via HAProxy, authentication for APIs, etc. because HTTP is well supported.

=item Access control list for methods of redis ;

=item Support to run as daemon (simple web-server), CGI, FastCGI or PSGI ;

=item HTTP Basic Auth security

=item Support for GET and POST.

=item Support for JSONP.

=item HTTP 1.1 pipelining (fastcgi and psgi)

=item CIDR authentication

=item Bad support for Websockets

=item Hooking, you can alter the behavior of an request method.

=item L<ReRe::Client> for write client applications in Perl.

=item REST interface to make your life easy in some world ;

=item Simple to install and use

=back

=head2 What is not ReRe ?

Please, don't use this application if ...

=over 4

=item If you are looking for a local database for cache or key-value

=item If you are looking for performance.

=back

=head1 Configuration

=head2 users.conf

    <users>
        <userro>
            password userro
            roles get
        </userro>

        <userrw>
            password userrw
            roles get set info
        </userrw>

        <userall>
            allow 127.0.0.1
            password userall
            roles all
        </userall>

        <userlocalnet>
            allow 192.168.0.0/24
            roles get info
        </userlocalnet>
    </users>

=head2 server.conf

    <server>
        host 127.0.0.1
        port 6379
        hooks Log
    </server>

=head1 Hooks

You can alter the behaivor of any method or create your own.

=head2 Example

    package ReRe::Hook::MyHook;
    use Moose::Role;

    sub _hook {
        my $self = shift;

        my $method = $self->method;
        my $args = $self->args;
        my $conn = $self->conn;

        if ( $method eq 'set' ) {
            my ( $key, $value ) = @{$args};

            if ( $key eq 'immutable' ) {
                return { err => 'this object is immutable' }
            }

            if ( $key eq 'semaphoro' ) {
                my $new_value = 0;
                $new_value = 1 if $value eq 'red';
                return $self->conn->set($key, $new_value)
            }
        }

    return 0; # if you want to process the origin method.
    }

    1;

=head1 Documentation

More information, you can read in L<http://www.rere.com.br>.

=head1 Development

ReRe is a open source project for everyone to participate. The code repository
is located on github. Feel free to send a bug report, a pull request, or a
beer.

L<http://www.github.com/maluco/ReRe>

=head1 METHODS

=head2 start

Start ReRe.

=head2 process

Process the request to redis server.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

perldoc ReRe
perldoc ReRe::Config
perldoc ReRe::Server
perldoc ReRe::User

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=ReRe>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/ReRe>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/ReRe>

=item * Search CPAN

L<http://search.cpan.org/dist/ReRe>

=back

=head1 AUTHOR

Thiago Rondon <thiago@nsms.com.br>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Thiago Rondon.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

